server/src/main/java/org/elasticsearch/rest/RestResponse.java

```diff
@@ -112,14 +112,28 @@ public RestResponse(RestChannel channel, RestStatus status, Exception e) throws
         this.status = status;
         ToXContent.Params params = channel.request();
         if (e != null) {
-            Supplier<?> messageSupplier = () -> String.format(
-                Locale.ROOT,
-                "path: %s, params: %s, status: %d",
-                channel.request().rawPath(),
-                channel.request().params(),
-                status.getStatus()
-            );
-            if (status.getStatus() < 500) {
+            Supplier<?> messageSupplier;
+            boolean nodeOrShardUnavailableException = ExceptionsHelper.isNodeOrShardUnavailableTypeException(e);
+            if (nodeOrShardUnavailableException) {
+                messageSupplier = () -> String.format(
+                    Locale.ROOT,
+                    "path: %s, params: %s, status: %d; error: %s %s",
+                    channel.request().rawPath(),
+                    channel.request().params(),
+                    status.getStatus(),
+                    e.getClass(),
+                    e.getMessage()
+                );
+            } else {
+                messageSupplier = () -> String.format(
+                    Locale.ROOT,
+                    "path: %s, params: %s, status: %d",
+                    channel.request().rawPath(),
+                    channel.request().params(),
+                    status.getStatus()
+                );
+            }
+            if (status.getStatus() < 500 || nodeOrShardUnavailableException) {
```

Comment:
For handling "node/shard unavailable" exceptions. Options here are:

1) do not log these at all
2) log at debug level with stacktrace
3) log at warn level without stacktrace
4) log at debug level without stacktrace

I chose number 4, since we already filter these out of our serverless dashboards, but let me know if you think we should pick a different option.

Reply:
++ another option would be to add a counter metric that keeps track of how many of these happen, for monitoring purposes. I think it's important to hear from Core/Infra what their view on these exception, whether they are useful logged, etc.

Reply:
I would prefer debug with stacktrace. If we are enabling debug logging, we likely want the level of granularity the stacktrace provides. Additionally, the code I see here should still provide the stacktrace since the exception is still passed to the logger, so I think what is implemented in this PR is actually 2.